### DISCUSSION TITLE
Streamlink not working on some HLS, but the web browser works.

### DISCUSSION THREAD

**Discussion by zydjohnHotmail** [streamlink/streamlink#4317] [Q&A] [answered]
### Checklist

- [X] This is a bug report and not a different kind of issue
- [X] [I have read the contribution guidelines](https://github.com/streamlink/streamlink/blob/master/CONTRIBUTING.md#contributing-to-streamlink)
- [X] [I have checked the list of open and recently closed bug reports](https://github.com/streamlink/streamlink/issues?q=is%3Aissue+label%3A%22bug%22)
- [X] [I have checked the commit log of the master branch](https://github.com/streamlink/streamlink/commits/master)

### Streamlink version

Latest stable release

### Description

Hello:
I have installed streamlink latest version (3.1.1) on Windows 10, and I want to download some HLS to save in local hard drive, sometimes, it works well, but sometimes, it didn’t work.
I have one specific example, in which, I can view the live HLS via Google Chrome browser, it uses JavaScript video player, but I inspected the network traffic, and I got the HLS M3U8 link, so I tried to download it and save to my PC hard drive, but it didn’t work.
So, I tried to debug to find out what went wrong, here is the detailed debug output:

### Debug log

```text
D:\Videos\Download>"C:\Program Files\Streamlink\Bin\Streamlink.exe" --stream-segment-attempts 200 --stream-segment-timeout 900.0 --http-header "Referer=https://live-streamfootball.com/"  --hls-duration 00:00:3480 https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8 best -f -o Test.mp4 -l debug
[cli][debug] OS:         Windows 10
[cli][debug] Python:     3.9.8
[cli][debug] Streamlink: 3.1.1
[cli][debug] Requests(2.26.0), Socks(1.7.1), Websocket(1.2.1)
[cli][debug] Arguments:
[cli][debug]  url=https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8
[cli][debug]  stream=['best']
[cli][debug]  --loglevel=debug
[cli][debug]  --output=9478_A.mp4
[cli][debug]  --force=True
[cli][debug]  --stream-segment-attempts=200
[cli][debug]  --stream-segment-timeout=900.0
[cli][debug]  --hls-duration=3480
[cli][debug]  --hls-segment-threads=10
[cli][debug]  --ffmpeg-ffmpeg=C:\Program Files\StreamLink\ffmpeg\ffmpeg.exe
[cli][debug]  --http-header=[('Referer', 'https://live-streamfootball.com/')]
[cli][info] Found matching plugin hls for URL https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8
[plugins.hls][debug] URL=https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8; params={}
[utils.l10n][debug] Language code: en_US
[cli][info] Available streams: live (worst, best)
[cli][info] Opening stream: live (hls)
[cli][debug] Checking file output
[stream.hls][debug] Reloading playlist
[cli][debug] Pre-buffering 8192 bytes
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.hls][debug] Reloading playlist
[stream.segmented][debug] Closing worker thread
[stream.segmented][debug] Closing writer thread
[stream.segmented][debug] Closing writer thread
[cli][error] Try 1/1: Could not open stream <HLSStream('https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8', None)> (Failed to read data from stream: Read timeout)
error: Could not open stream <HLSStream('https://m3u.1proxy.xyz/sv2/go_13/bingsport_gcore.m3u8', None)>, tried 1 times, exiting
[cli][info] Closing currently open stream…
```

**Comment by mkbloke** [comment]
Maybe you're missing an/some HTTP header(s)?

**Comment by bastimeyer** [comment, marked as answer]
> Failed to read data from stream: Read timeout

You have run into the default `--stream-timeout` value of 60 seconds. If there is no output during this time period, streamlink will stop.
https://streamlink.github.io/cli.html#cmdoption-stream-timeout

As you can see in the debug log, it is not queuing any segments, which means you can't access those segments. You have for some very weird reason set the re-attempt number to 200 and the segment timeout number to 900 seconds. This means that once a segment fails, eg. due to bad permissions on the server due to lack of proper authentication, etc., Streamlink will re-attempt it in 900 seconds, for at most 200 times. Complete non-sense. But since there's no stream output after 60 seconds, the process exits with a timeout error.

Remove `--stream-segment-attempts` and `--stream-segment-timeout` and add the necessary HTTP headers for being able to access those segments.

**Reply by zydjohnHotmail** [reply]
Hello:
I can delete those incorrect parameters, but how I can figure out which HTTP headers to add?
Please advise,
By the way: the following is a HTTP request headers for a similar HLS stream:
Request Header
:authority: m3u.1proxy.xyz
:method: GET
:path: /sv3/go_29/index_gcore.m3u8
:scheme: https
accept: */*
accept-encoding: gzip, deflate, br
accept-language: en-GB,en-US;q=0.9,en;q=0.8
if-none-match: W/"215-p2UTSyZKoY5ZyDFI0ahqA3FM9YQ"
origin: https://live-streamfootball.com
referer: https://live-streamfootball.com/
sec-ch-ua: " Not;A Brand";v="99", "Google Chrome";v="97", "Chromium";v="97"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
sec-fetch-dest: empty
sec-fetch-mode: cors
sec-fetch-site: cross-site
user-agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.99 Safari/537.36
I can provide more information, if you find more clues.

**Reply by gravyboat** [reply]
This is not technical support for which headers need to be added so random websites work, you'll have to figure that out on your own.
